client/protocol.go

```diff
@@ -0,0 +1,413 @@
+					fetchClusterRespCh <- nil
+				} else {
+					fetchClusterRespCh <- res
+				}
+			}()
+		}
+
+		for {
+			fetchClusterRes := <-fetchClusterRespCh
+
+			if fetchClusterRes == nil {
+				continue
+			}
+
+			if maxTerm < fetchClusterRes.Term {
+				maxTerm = fetchClusterRes.Term
+				leader = fetchClusterRes.LeaderId
+				okCnt = 1
+			} else if maxTerm == fetchClusterRes.Term {
+				leader = fetchClusterRes.LeaderId
+				okCnt += 1
+			}
+			if okCnt >= majorityCnt {
+				break
+			}
+		}
+
+		return leader
+	}
+}
+
+type ProposeResponse struct {
+	SyncResp           *xlineapi.SyncResponse
+	CommandResp        *xlineapi.CommandResponse
+	IsFastRoundSuccess bool
+}
+
+// Curp client settings
+type clientTimeout struct {
+	// Curp client wait idle
+	idleTimeout time.Duration
+	// Curp client wait sync timeout
+	waitSyncedTimeout time.Duration
+	// Curp client propose request timeout
+	proposeTimeout time.Duration
+	// Curp client retry interval
+	retry_timeout time.Duration
+}
+
+func NewDefaultClientTimeout() clientTimeout {
+	return clientTimeout{
+		idleTimeout:       1 * time.Second,
+		waitSyncedTimeout: 2 * time.Second,
+		proposeTimeout:    1 * time.Second,
+		retry_timeout:     50 * time.Millisecond,
+	}
+}
+
+type CommandSyncError struct {
+	*curpapi.CommandSyncError
+}
+
+type ProposeError struct {
+	*xlineapi.ExecuteError
+}
+
+func (e CommandSyncError) Error() string {
+	switch cmdSyncErr := e.CommandSyncError.CommandSyncError.(type) {
+	case *curpapi.CommandSyncError_WaitSync:
+		switch syncErr := cmdSyncErr.WaitSync.WaitSyncError.(type) {
+		case *curpapi.WaitSyncError_Redirect:
+			return fmt.Sprintf("Sync redirect error. SeverId: %v, Term: %v\n", syncErr.Redirect.ServerId, syncErr.Redirect.Term)
+		case *curpapi.WaitSyncError_Other:
+			return fmt.Sprintln("Sync other error.", syncErr.Other)
+		default:
+			return "Sync error"
+		}
+	case *curpapi.CommandSyncError_Execute:
+		var execute xlineapi.ExecuteError
+		err := proto.Unmarshal(cmdSyncErr.Execute, &execute)
+		if err != nil {
+			logrus.Warn(err)
+		}
+		return fmt.Sprintf("Command sync execute error. %+v", &execute)
+	case *curpapi.CommandSyncError_AfterSync:
+		var afterSync xlineapi.ExecuteError
+		err := proto.Unmarshal(cmdSyncErr.AfterSync, &afterSync)
+		if err != nil {
+			logrus.Warn(err)
+		}
+		return fmt.Sprintf("Command after sync execute error. %+v", &afterSync)
+	default:
+		return "Command sync error"
+	}
+}
+
+func (e ProposeError) Error() string {
+	return fmt.Sprintf("Execute error %+v\n", e.ExecuteError.Error)
+}
+
+// Get the superquorum for curp protocol
+// Although curp can proceed with f + 1 available replicas, it needs f + 1 + (f + 1)/2 replicas
+// (for superquorum of witnesses) to use 1 RTT operations. With less than superquorum replicas,
+// clients must ask masters to commit operations in f + 1 replicas before returning result.(2 RTTs).
+func superQuorum(nodes int) int {
+	faultTolerance := nodes / 2
+	return (faultTolerance+1)/2 + 1
```

Comment:
This should be `faultTolerance + 1 + (faultTolerance + 1) / 2`

Reply:
done